### ISSUE TITLE
adaptive loss or class weights

### ISSUE THREAD

**Issue by faltinl** [rstudio/keras#529]
When treating instationary, albeit statistically slowly changing data and the network is learning faster than the data is changing, it would be interesting to adapt the network on the fly as these changes occur. Just imagine the case that in a classification task the relative proportion of the classes changes; or in the case of an inception network the contributions of/to the various channels might change and concatenation of these channels should be adapted accordingly in order to keep the network optimized.

My question is: How would it be possible to achieve this aim?

From Keras documentation I cannot find out when exactly the parameters class_weight (for multiple class problems) and loss_weight (for multiple output problems) are read by the program and applied to the network. 

1. Are they only read at compilation?  
2. Are they read between single fit()-cycles, i.e. after a preprogrammed number of epochs?
3. Are they read between single epochs of one particular fit()-cycle?

Obviously, case 3. would be the one providing a maximum of programming freedom. In contrast to that, case 1. would require new compilation, which probably means that network parameters learned until then will be lost. 

Case 2. could potentially indicate a solution - if the commands save/load_model_weights_hdf5 really concern _only_ the network layers and not loss/class_weights: one could then save the newly learned model layers, load a new list of class/loss_weights (e.g. by compilation - or how else?), and then reload the model_weights. Again, the documentation does not provide any hint to that approach.

Any pertinent comment would be welcome - thanks in advance!
Leo

**Comment by faltinl** [comment]
Hi,-
and thank you for this excellent demo example! True, class_weights are read only at starting up the `fit()`-process and can therefore be changed to be read anew upon every re-start of `fit()`. I even think that changing the `class_weights` and reading the list could be done without removing/recreating the model and reloading the saved layer weights, as these steps don't change anything after all. Referring back to my initial question, this in fact corresponds to case no. 2, where new class_weights can be read between `fit()`-cycles.

Thus remains the problem of how to adapt a given model to modified `loss_weights` in case of multiple outputs (see [https://keras.rstudio.com/articles/functional_api.html](url) for a simple well documented example). They are defined among other parameters within the command 

```
compile(                   
   optimizer(...), 
   loss = list(output1 = ..., output2 = ..., ...),
   loss_weights = list(output1 = ..., output2 = ..., ...),
   metrics = list(output1 = ..., output2 = ..., ...)
) 
```
and I don't know whether this command may be used without destoying an existing configuration of layer-weights. On the other hand, applying `compile(...)` _before_ loading previously saved layer-weights might possibly destroy the `compile(...)`-parameters as soon as the layer-weights are loaded again.

**Comment by skeydan** [comment]
That's why I included saving the model weights and restoring them in the above test code.
You can see that after restore, training does not resume from scratch.

So you can take that test code and adapt it, to check what happens in the case of modifying loss weights.

**Comment by faltinl** [comment]
Yepp - however, the test code doesn't have multiple outputs. So I took one existing DL program to carry out this test and really: running the `compile()`-command didn't change the layer weights learned before. That's great - I always assumed the worst!-)

So, both `class_weights` as well as `loss_weights` may be changed in their respective places simply after interrupting the learning process; and learning will then resume based on the new lists.

Thanks a lot for your support - that was very helpful indeed!
Rgds
Leo

**Comment by skeydan** [comment]
Great, thanks for reporting the result!
